File: easy/canContruct.js
/**
 * @param {string} ransomNote
 * @param {string} magazine
 * @return {boolean}
 */
const canConstruct = (ransomNote, magazine) => {
  const magazineCount = {};

  for (const char of magazine) {
    magazineCount[char] = (magazineCount[char] || 0) + 1;
  }

  for (const char of ransomNote) {
    if (!magazineCount[char]) {
      return false;
    }
    magazineCount[char]--;
  }

  for (const char in magazineCount) {
    if (magazineCount[char] < 0) {
      return false;
    }
  }

  return true;
};

File: easy/canContruct.java
import java.util.HashMap;
import java.util.Map;

class Solution {

    public boolean canConstruct(String ransomNote, String magazine) {
        Map<String, Integer> magazineHashMap = new HashMap<>();
        
        for(char c : magazine.toCharArray()) {
            String key = String.valueOf(c);
            magazineHashMap.put(key, magazineHashMap.getOrDefault(key, 0) + 1);
        }

        for(char c : ransomNote.toCharArray()) {
            String key = String.valueOf(c);
            if(!magazineHashMap.containsKey(key) || magazineHashMap.get(key) == 0) {
                return false;
            }
            magazineHashMap.put(key, magazineHashMap.get(key) - 1);
        } 

        return true; 
    }
}
